File: ts/src/scope.test.ts
import { test } from "node:test";
import { strict as assert } from "node:assert";
import type { Claims } from "./claims.js";
import {
  scopesFrom,
  scopeAllows,
  scopeAllowsAny,
  decideScope,
  validateScopePolicy,
  createScopeMiddleware,
  fastifyScopeHook,
  type ScopePolicy,
  type ScopeDecision,
} from "./scope.js";

function claimsWith(scope: unknown): Claims {
  return { iss: "i", sub: "s", aud: "a", exp: 0, iat: 0, scope } as unknown as Claims;
}

// Every way the predicate can be asked a question it cannot answer. Each must
// be false.
//
// The empty-required case is the one worth arguing about: scopeAllows(c) with
// no scopes returns FALSE, not true. Vacuous-true on an empty requirement is
// how a gate silently stops gating — a route someone forgot to configure would
// pass every caller — and a genuinely public route is DECLARED, not inferred.
test("scopeAllows fails closed", () => {
  const full = claimsWith("a b c");
  const cases: Array<[string, Claims | null, string[], boolean]> = [
    ["null claims", null, ["a"], false],
    ["no scope claim", claimsWith(undefined), ["a"], false],
    ["empty scope claim", claimsWith(""), ["a"], false],
    ["whitespace-only", claimsWith("   "), ["a"], false],
    ["no required scopes is NOT vacuously true", full, [], false],
    ["single hit", full, ["b"], true],
    ["all-of, all present", full, ["a", "c"], true],
    ["all-of, one missing", full, ["a", "z"], false],
    ["no prefix implication", claimsWith("read"), ["read:orders"], false],
    ["no suffix implication", claimsWith("read:orders"), ["read"], false],
    ["no wildcard expansion", claimsWith("*"), ["anything"], false],
    ["case-sensitive", claimsWith("read"), ["Read"], false],
    // RFC 9068 §2.2.3 makes this claim a STRING. Quietly accepting an array
    // would mask a wire mismatch that ought to be loud.
    ["array-shaped claim is not read", claimsWith(["a"]), ["a"], false],
    ["number-shaped claim is not read", claimsWith(7), ["7"], false],
  ];
  for (const [name, claims, required, want] of cases) {
    assert.equal(scopeAllows(claims, ...required), want, name);
  }
});

test("scopeAllowsAny differs from scopeAllows and fails closed", () => {
  const c = claimsWith("a b");
  assert.equal(scopeAllowsAny(c, "z", "b"), true, "one hit is enough");
  assert.equal(scopeAllowsAny(c, "y", "z"), false, "no hits");
  assert.equal(scopeAllowsAny(c), false, "empty required is not vacuously true");
  assert.equal(scopeAllowsAny(null, "a"), false, "null claims");
  // If these two ever agree on a partially-held set, one of them is decoration.
  assert.notEqual(scopeAllows(c, "a", "z"), scopeAllowsAny(c, "a", "z"));
});

test("scopesFrom preserves the issuer's order and splits on runs", () => {
  assert.deepEqual(scopesFrom(claimsWith("c  a   b")), ["c", "a", "b"]);
  assert.deepEqual(scopesFrom(null), []);
  assert.deepEqual(scopesFrom(claimsWith(["a"])), []);
});

// The property that makes the route map safe: adding an endpoint and forgetting
// to declare its scope produces a LOCKED DOOR, not an open one.
test("a policy denies by default", () => {
  const policy: ScopePolicy = [{ path: "/orders/**", scopes: ["orders:read"] }];
  const c = claimsWith("orders:read admin");

  // PRECONDITION: the declared route IS allowed, so the denial below is
  // attributable to the missing declaration and not to a broken policy.
  assert.equal(decideScope(policy, c, "GET", "/orders/42").allowed, true,
    "PRECONDITION: the declared route must be allowed");

  const d = decideScope(policy, c, "GET", "/invoices/42");
  assert.equal(d.allowed, false, "an UNDECLARED route must be denied, even to a token holding every scope");
  assert.equal(d.matched, false, "matched=false lets a caller tell a config gap from an authz failure");

  // A missing policy denies too — a wiring mistake must not look like a
  // deliberately open service.
  assert.equal(decideScope(null, c, "GET", "/orders/42").allowed, false);
  assert.equal(decideScope(undefined, c, "GET", "/orders/42").allowed, false);
});

test("public, anyOf, method and first-match-wins", () => {
  const policy: ScopePolicy = [
    { path: "/health", public: true },
    { path: "/orders/*/export", scopes: ["orders:export"] },
    { path: "/orders/**", method: "GET", scopes: ["orders:read"] },
    { path: "/orders/**", scopes: ["orders:write", "orders:read"] },
    { path: "/reports/**", scopes: ["r:a", "r:b"], anyOf: true },
  ];

  assert.equal(decideScope(policy, null, "GET", "/health").allowed, true,
    "a public route must allow a request with no claims at all");

  const read = claimsWith("orders:read");
  assert.equal(decideScope(policy, read, "GET", "/orders/7").allowed, true);
  assert.equal(decideScope(policy, read, "POST", "/orders/7").allowed, false,
    "POST falls through to the write rule");

  // /orders/7/export matches BOTH the export rule and the GET rule; the export
  // rule is first, so it decides.
  const exp = decideScope(policy, read, "GET", "/orders/7/export");
  assert.equal(exp.allowed, false, "the earlier, more specific rule must decide");
  assert.deepEqual(exp.required, ["orders:export"]);

  assert.equal(decideScope(policy, read, "POST", "/orders/7").allowed, false,
    "a two-scope rule requires BOTH by default");
  assert.equal(decideScope(policy, claimsWith("r:a"), "GET", "/reports/x").allowed, true,
    "anyOf passes on one of two");

  assert.deepEqual(decideScope(policy, read, "POST", "/orders/7").missing, ["orders:write"]);
  assert.deepEqual(decideScope(policy, claimsWith("nope"), "GET", "/reports/x").missing, [],
    "an anyOf denial has no single missing scope");
});

// These are the mistakes a partner makes once and should learn about at
// startup, not from traffic.
test("validateScopePolicy reports every problem, not the first", () => {
  const errs = validateScopePolicy([
    { path: "", scopes: ["a"] },
    { path: "/a", public: true, scopes: ["a"] },
    { path: "/b" },
    { path: "/c", scopes: ["has space"] },
    { path: "/ok", scopes: ["fine"] },
  ]);
  assert.equal(errs.length, 4, `want 4 errors, got ${JSON.stringify(errs)}`);
  for (const e of errs) assert.ok(e.message.length > 0);
  assert.deepEqual(validateScopePolicy([{ path: "/ok", scopes: ["fine"] }]), []);
});

class Res {
  statusCode = 200;
  headers: Record<string, string> = {};
  body = "";
  setHeader(n: string, v: string) { this.headers[n.toLowerCase()] = v; }
  end(chunk?: string) { if (chunk) this.body += chunk; }
}

// Telling an unauthorized caller which permissions they lack hands out a map of
// the API's authority model for free. The names go to the SERVER, via
// onScopeDenied.
test("the Express adapter 403s without leaking the scope names", () => {
  const policy: ScopePolicy = [{ path: "/secret", scopes: ["very:secret:permission"] }];
  let seen: ScopeDecision | null = null;
  const mw = createScopeMiddleware(policy, { onScopeDenied: (_r, d) => { seen = d; } });

  const res = new Res();
  let nexted = false;
  mw({ method: "GET", url: "/secret?q=1" }, res, () => { nexted = true; });

  assert.equal(nexted, false, "a denied request must not reach the handler");
  assert.equal(res.statusCode, 403);
  assert.ok(res.body.includes("insufficient_scope"), "RFC 6750 §3.1 code");
  assert.ok(!res.body.includes("very:secret:permission"), "the 403 body LEAKED the required scope name");
  assert.ok(seen, "onScopeDenied must fire, or the names are lost to the server too");
  assert.deepEqual(seen!.missing, ["very:secret:permission"]);

  // PRECONDITION for the negative above: with the scope, the handler runs.
  // Without this, an adapter that refused everything would pass every assertion.
  let allowed = false;
  mw(
    { method: "GET", url: "/secret", realmid: claimsWith("very:secret:permission") },
    new Res(),
    () => { allowed = true; },
  );
  assert.equal(allowed, true, "an authorized request must reach the handler");
});

test("the Fastify hook enforces the same policy", () => {
  const policy: ScopePolicy = [{ path: "/secret", scopes: ["s"] }];
  const hook = fastifyScopeHook(policy);

  let status = 0;
  let body: unknown = null;
  let done = false;
  hook({ method: "GET", url: "/secret" }, {
    code(s: number) { status = s; return { send(b: unknown) { body = b; return null; } }; },
  }, () => { done = true; });
  assert.equal(status, 403);
  assert.equal(done, false);
  assert.equal((body as { error: { code: string } }).error.code, "insufficient_scope");

  let allowed = false;
  hook({ method: "GET", url: "/secret", realmid: claimsWith("s") }, {
    code() { throw new Error("must not reply"); },
  }, () => { allowed = true; });
  assert.equal(allowed, true);
});

File: go/scope_test.go
package realmid

import (
	"context"
	"net/http"
	"net/http/httptest"
	"reflect"
	"testing"
)

func claimsWithScope(s string) *Claims {
	return &Claims{Extra: map[string]any{"scope": s}}
}

// TestScopeAllows_FailsClosed walks every way the predicate can be asked a
// question it cannot answer. Each one must be false.
//
// The empty-required case is the one worth arguing about: ScopeAllows(c) with
// no scopes returns FALSE, not true. Vacuous-true on an empty requirement is
// how a gate silently stops gating — a route someone forgot to configure would
// pass every caller — and a genuinely public route is declared, not inferred.
func TestScopeAllows_FailsClosed(t *testing.T) {
	full := claimsWithScope("a b c")
	cases := []struct {
		name     string
		claims   *Claims
		required []string
		want     bool
	}{
		{"nil claims", nil, []string{"a"}, false},
		{"no scope claim", &Claims{}, []string{"a"}, false},
		{"empty scope claim", claimsWithScope(""), []string{"a"}, false},
		{"whitespace-only scope claim", claimsWithScope("   "), []string{"a"}, false},
		{"no required scopes is NOT vacuously true", full, nil, false},
		{"single hit", full, []string{"b"}, true},
		{"all-of, all present", full, []string{"a", "c"}, true},
		{"all-of, one missing", full, []string{"a", "z"}, false},
		// No pattern matching, exactly as CapAllows states it.
		{"no prefix implication", claimsWithScope("read"), []string{"read:orders"}, false},
		{"no suffix implication", claimsWithScope("read:orders"), []string{"read"}, false},
		{"no wildcard expansion", claimsWithScope("*"), []string{"anything"}, false},
		{"case-sensitive", claimsWithScope("read"), []string{"Read"}, false},
		// The claim is a STRING, not an array (RFC 9068 §2.2.3).
		{"array-shaped claim is not read", &Claims{Extra: map[string]any{"scope": []any{"a"}}}, []string{"a"}, false},
	}
	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			if got := ScopeAllows(tc.claims, tc.required...); got != tc.want {
				t.Errorf("ScopeAllows(%v) = %v, want %v", tc.required, got, tc.want)
			}
		})
	}
}

func TestScopeAllowsAny(t *testing.T) {
	c := claimsWithScope("a b")
	if !ScopeAllowsAny(c, "z", "b") {
		t.Error("any-of must pass on a single hit")
	}
	if ScopeAllowsAny(c, "y", "z") {
		t.Error("any-of must fail when none hit")
	}
	if ScopeAllowsAny(c) {
		t.Error("any-of with no required scopes must be false, not vacuously true")
	}
	if ScopeAllowsAny(nil, "a") {
		t.Error("any-of must fail closed on nil claims")
	}
	// The two predicates must genuinely differ, or one of them is decoration.
	if ScopeAllows(c, "a", "z") == ScopeAllowsAny(c, "a", "z") {
		t.Error("all-of and any-of agree on a partially-held set; one of them is not doing its job")
	}
}

func TestScopesFrom_PreservesOrderAndSplitsOnRuns(t *testing.T) {
	got := ScopesFrom(claimsWithScope("c  a   b"))
	if !reflect.DeepEqual(got, []string{"c", "a", "b"}) {
		t.Errorf("ScopesFrom = %v, want the issuer's order preserved", got)
	}
	if ScopesFrom(nil) != nil {
		t.Error("ScopesFrom(nil) must be nil")
	}
}

// TestScopePolicy_DeniesByDefault is the property that makes the route map
// safe: adding an endpoint and forgetting to declare its scope produces a
// LOCKED DOOR, not an open one.
func TestScopePolicy_DeniesByDefault(t *testing.T) {
	p := ScopePolicy{Rules: []ScopeRule{
		{Path: "/orders/**", Scopes: []string{"orders:read"}},
	}}.Compile()
	c := claimsWithScope("orders:read admin")

	// PRECONDITION: the declared route is allowed, so the denial below is
	// attributable to the missing declaration and not to a broken policy.
	if !p.Decide(c, "GET", "/orders/42").Allowed {
		t.Fatal("PRECONDITION FAILED — the declared route must be allowed")
	}

	d := p.Decide(c, "GET", "/invoices/42")
	if d.Allowed {
		t.Error("an UNDECLARED route must be denied, even to a token carrying every scope")
	}
	if d.Matched {
		t.Error("Matched must be false so a caller can tell a config gap from an authz failure")
	}

	// A nil policy denies too — a wiring mistake must not look like a
	// deliberately open service.
	var nilPolicy *CompiledScopePolicy
	if nilPolicy.Decide(c, "GET", "/orders/42").Allowed {
		t.Error("a nil policy must deny")
	}
}

func TestScopePolicy_PublicAnyOfMethodAndOrder(t *testing.T) {
	p := ScopePolicy{Rules: []ScopeRule{
		{Path: "/health", Public: true},
		{Path: "/orders/*/export", Scopes: []string{"orders:export"}},
		{Path: "/orders/**", Method: "GET", Scopes: []string{"orders:read"}},
		{Path: "/orders/**", Scopes: []string{"orders:write", "orders:read"}},
		{Path: "/reports/**", Scopes: []string{"r:a", "r:b"}, AnyOf: true},
	}}.Compile()

	t.Run("public needs no token at all", func(t *testing.T) {
		if !p.Decide(nil, "GET", "/health").Allowed {
			t.Error("a Public route must allow a request with no claims")
		}
	})
	t.Run("method narrows", func(t *testing.T) {
		read := claimsWithScope("orders:read")
		if !p.Decide(read, "GET", "/orders/7").Allowed {
			t.Error("GET must match the read rule")
		}
		if p.Decide(read, "POST", "/orders/7").Allowed {
			t.Error("POST must fall through to the write rule and be denied")
		}
	})
	t.Run("first match wins, so a specific rule placed first narrows", func(t *testing.T) {
		// /orders/7/export matches BOTH the export rule and the GET rule. The
		// export rule is first, so it decides — a token holding only orders:read
		// must be refused there.
		read := claimsWithScope("orders:read")
		d := p.Decide(read, "GET", "/orders/7/export")
		if d.Allowed {
			t.Error("the earlier, more specific rule must decide")
		}
		if !reflect.DeepEqual(d.Required, []string{"orders:export"}) {
			t.Errorf("the export rule should have matched, got Required=%v", d.Required)
		}
	})
	t.Run("all-of is the default and any-of must be asked for", func(t *testing.T) {
		half := claimsWithScope("orders:read")
		if p.Decide(half, "POST", "/orders/7").Allowed {
			t.Error("a two-scope rule must require BOTH by default")
		}
		if !p.Decide(claimsWithScope("r:a"), "GET", "/reports/x").Allowed {
			t.Error("AnyOf must pass on one of two")
		}
	})
	t.Run("Missing names the gap, and only on an all-of denial", func(t *testing.T) {
		d := p.Decide(claimsWithScope("orders:read"), "POST", "/orders/7")
		if !reflect.DeepEqual(d.Missing, []string{"orders:write"}) {
			t.Errorf("Missing = %v, want [orders:write]", d.Missing)
		}
		if any := p.Decide(claimsWithScope("nope"), "GET", "/reports/x"); len(any.Missing) != 0 {
			t.Errorf("an AnyOf denial has no single missing scope; got %v", any.Missing)
		}
	})
}

// TestScopePolicy_ValidateCatchesConfigErrors — these are the mistakes a
// partner makes once and should learn about at startup, not from traffic.
func TestScopePolicy_ValidateCatchesConfigErrors(t *testing.T) {
	errs := ScopePolicy{Rules: []ScopeRule{
		{Path: "", Scopes: []string{"a"}},
		{Path: "/a", Public: true, Scopes: []string{"a"}},
		{Path: "/b"},
		{Path: "/c", Scopes: []string{"has space"}},
		{Path: "/ok", Scopes: []string{"fine"}},
	}}.Validate()
	if len(errs) != 4 {
		t.Fatalf("want 4 errors (empty path, public-with-scopes, no-scopes-not-public, bad charset), got %d: %v",
			len(errs), errs)
	}
	// Every error names its rule, or a partner with a 40-route map cannot find it.
	for _, e := range errs {
		ce, ok := e.(*ScopeConfigError)
		if !ok {
			t.Fatalf("want *ScopeConfigError, got %T", e)
		}
		if ce.Error() == "" {
			t.Error("a config error must render")
		}
	}
	if errs := (ScopePolicy{Rules: []ScopeRule{{Path: "/ok", Scopes: []string{"fine"}}}}).Validate(); len(errs) != 0 {
		t.Errorf("a valid policy must produce no errors, got %v", errs)
	}
}

// TestScopeMiddleware_403AndDoesNotLeakTheScopeNames.
//
// Telling an unauthorized caller which permissions they lack hands out a map of
// the API's authority model for free. The names go to the SERVER, through
// OnScopeDenied.
func TestScopeMiddleware_403AndDoesNotLeakTheScopeNames(t *testing.T) {
	p := ScopePolicy{Rules: []ScopeRule{
		{Path: "/secret", Scopes: []string{"very:secret:permission"}},
	}}.Compile()

	var seen ScopeDecision
	var called bool
	h := p.Middleware(ScopeMiddlewareOptions{
		OnScopeDenied: func(_ *http.Request, d ScopeDecision) { seen, called = d, true },
	})(http.HandlerFunc(func(w http.ResponseWriter, _ *http.Request) {
		w.WriteHeader(http.StatusTeapot)
	}))

	// Denied: no claims on the context at all.
	rec := httptest.NewRecorder()
	h.ServeHTTP(rec, httptest.NewRequest("GET", "/secret", nil))
	if rec.Code != http.StatusForbidden {
		t.Fatalf("status = %d, want 403", rec.Code)
	}
	if body := rec.Body.String(); !contains(body, "insufficient_scope") {
		t.Errorf("body should carry the RFC 6750 §3.1 code, got %s", body)
	}
	if body := rec.Body.String(); contains(body, "very:secret:permission") {
		t.Errorf("the 403 body LEAKED the required scope name: %s", body)
	}
	if !called {
		t.Fatal("OnScopeDenied must fire, or the names are lost to the server too")
	}
	if len(seen.Missing) != 1 || seen.Missing[0] != "very:secret:permission" {
		t.Errorf("the server-side hook must carry the missing names, got %v", seen.Missing)
	}

	// PRECONDITION for the negative above: with the scope, the handler runs.
	// Without this, a middleware that refused everything would pass every
	// assertion so far.
	req := httptest.NewRequest("GET", "/secret", nil)
	req = req.WithContext(context.WithValue(req.Context(), claimsKey, claimsWithScope("very:secret:permission")))
	rec2 := httptest.NewRecorder()
	h.ServeHTTP(rec2, req)
	if rec2.Code != http.StatusTeapot {
		t.Fatalf("an authorized request must reach the handler; status = %d", rec2.Code)
	}
}

func contains(s, sub string) bool {
	for i := 0; i+len(sub) <= len(s); i++ {
		if s[i:i+len(sub)] == sub {
			return true
		}
	}
	return false
}
